### ISSUE TITLE
MSDNS: Freezes after Remote Code Execution statement

### ISSUE THREAD

**Issue by JoeIzzard** [StackExchange/dnscontrol#2621]
**Describe the bug**
Running a preview against our MSDNS provider, I get the info message saying that powershell commands will run on the remote machine, then it does nothing. Trying again with `-v` I get all the messages of my `require` files being loaded, the INFO statement then nothing.

I don't think it's actually establishing a connection as giving it an incorrect password has the same effect.

**DNS Provider**
 - MSDNS

**Additional context**
Changing the Provider to Bind has it complete fairly quickly so I don't think it's an issue with my general set up. I have tried using just DNSServer option however it fails as it doesn't have the DNS Powershell commands.

Our MSDNS is on our Domain Controller. DNSControl is being run from a secondary (Non-DC) server that will act as a GitLab Runner eventually doing this automatically.

I am able to manually create a PSSession in PowerShell.

**Comment by tlimoncelli** [comment]
Hi there!

That's an interesting situation!

As the docs say "This is a new provider and has not been tested extensively, especially the pssession feature." [link](https://docs.dnscontrol.org/service-providers/providers/msdns).

To be blunt, I don't think the remote machine feature (pssession) works.  I've never had anyone confirm it was working for them.  It is a feature of the third-party "go-powershell" module we use, and to be honest I don't think even that author has ever seen it work. There are a few bug reports about it with no responses.  

We do use the MSDNS provider at Stack, but we use it on a Github runner which is domain joined.  It is not a DC.  We use the `dnsserver` setting to point to our DC, and set permissions on our zone to permit the user to do what they need to do.

When you say the remote commands do nothing... can you try waiting a VERY long time?  (like... overnight).  I ask because the provider can be slow for very large zones.   The powershell generates its results as JSON data.  PowerShell's JSON generator is slow. The JSON data is then read by DNSControl (I'm not going to claim that Go's JSON reader is very fast either!).  We made some optimizations 2 years ago (311a6d74c) to generate smaller JSON files (we were throwing away most of the data received).

I'd also suggest asking on the mailing list (https://groups.google.com/g/dnscontrol-discuss).  You'll reach a broader audience that way.

Lastly, if you are interested in contributing code, there are plenty of improvements that could be made with the MSDNS provider.

In summary:

* Try using dnsserver instead of pssession
* Try waiting a LONG time to see if the problem is the JSON generator
* Ask on the mailing list

Hope that helps!

Tom

**Comment by JoeIzzard** [comment]
The machine I am trying to run it from is Domain Joined, so I'll see if I can get it working without the PSSession stuff. 

Do you happen to know if you needed to install the DNS Server role on your runner to get the PowerShell CMDLets?

I did leave it running overnight to see if it was just the JSON generator taking a long time, with no luck. It was running for about 15 hours but made no further progress than any other attempt from what I can see. It is a larger zone, we have around 800 records total, but I think the PSSession stuff just isn't working.

I'll reach out on the mailing list too!

**Comment by tlimoncelli** [comment]
I propose we remove pssession from the docs.  I've never seen it work.  I've prepared https://github.com/StackExchange/dnscontrol/pull/2689 on the assumption you agree.

**Comment by JoeIzzard** [comment]
I'd agree, in the end I abandoned PSSession and instead used Windows Remote Server Administration Tools to update the Domain Controller from the Gitlab runner which was domain joined.

From the testing I have done so far that seems to be the solution although I haven't yet had the chance to try a push due to our change windows etc.
